Test/HigherOrderIntrinsicSpecification/ChecksAndUsabilityCoverage.dfy

```diff
@@ -0,0 +1,283 @@
+// RUN: %exits-with 4 %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+// Tests for well-formedness of FunctionCallExpr
+
+function MyFunction(x: int, S: set<object>): real
+  requires x < 100
+  reads S
+{
+  assert true;
+  (x + 12) as real
+}
+
+// In methods, there are no reads checks:
+
+method MCaller0(x: int, S: set<object>) {
+  var myFunctionRequires := MyFunction.requires(x, S);
+  var myFunctionReads := MyFunction.reads(x, S); // error: precondition (x < 100) may not hold
+}
+
+method MCaller1(S: set<object>) {
+  var myFunctionRequires := MyFunction.requires(5, S);
+  var myFunctionReads := MyFunction.reads(5, S);
+}
+
+// In functions, reads checks also matter:
+
+// .requires
+
+function FCallerRequires1(S: set<object>): real {
+  var myFunctionRequires := MyFunction.requires(5, S); // error: insufficient reads clause to call MyFunction.requires
+  3.14
+}
+
+function FCallerRequires2(S: set<object>): real
+  reads S
+{
+  var myFunctionRequires := MyFunction.requires(5, S);
+  assert myFunctionRequires;
+  3.14
+}
+
+function FCallerRequires3(S: set<object>): real
+  requires MyFunction.requires(5, S) // error: insufficient reads clause
+{
+  var myFunctionRequires := MyFunction.requires(5, S);
+  assert myFunctionRequires;
+  3.14
+}
+
+function FCallerRequires4(S: set<object>): real
+  requires MyFunction.requires(5, S)
+  reads S
+{
+  var myFunctionRequires := MyFunction.requires(5, S);
+  assert myFunctionRequires;
+  3.14
+}
+
+// .reads
+
+function FCallerReads0(x: int, S: set<object>): real {
+  var myFunctionReads := MyFunction.reads(x, S); // error (x2): precondition failure and insufficient reads clause
+  3.14
+}
+
+function FCallerReads1(S: set<object>): real {
+  var myFunctionReads := MyFunction.reads(5, S); // error: insufficient reads clause
+  3.14
+}
+
+function FCallerReads3(S: set<object>): real
+  reads S
+{
+  var myFunctionReads := MyFunction.reads(5, S);
+  3.14
+}
+
+function FCallerReads4(S: set<object>): real
+  requires MyFunction.requires(5, S)
+  reads MyFunction.reads(5, S)
+{
+  var myFunctionRequires := MyFunction.requires(5, S);
+  var myFunctionReads := MyFunction.reads(5, S);
+  assert myFunctionRequires;
+  assert myFunctionReads == S;
+  3.14
+}
+
+// Usability tests
+
+ghost function CallIfPossible0(x: int, S: set<object>): real {
+  if MyFunction.requires(x, S) then // error: insufficient reads clause to call MyFunction.requires
+    MyFunction(x, S)
+  else
+    2.7
+}
+
+ghost function CallIfPossible3(x: int, S: set<object>): real
+  requires x < 100
```

Comment:
Removing the precondition `requires x < 100` here shouldn't make any difference right?